newIDE/app/src/ObjectEditor/Editors/SpriteEditor/SpritesList.js

```diff
@@ -1,160 +1,239 @@
+
+const styles = {
+  spritesList: {
+    whiteSpace: 'nowrap',
+    overflowY: 'hidden',
+  },
+  thumbnailExtraStyle: {
+    marginRight: 10,
+  },
+  addSpriteButton: {
+    ...thumbnailContainerStyle,
+    background: '#FFF',
+  },
+  spriteThumbnailImage: {
+    maxWidth: SPRITE_SIZE,
+    maxHeight: SPRITE_SIZE,
+    verticalAlign: 'middle',
+  },
+};
+
+const AddSpriteButton = SortableElement(({ displayHint, onAdd, onEdit }) => {
+  return (
+    <div style={styles.addSpriteButton}>
+      <IconButton onClick={onAdd} style={styles.spriteThumbnailImage}>
+        <Add />
+      </IconButton>
+      <IconButton onClick={onEdit} style={styles.spriteThumbnailImage}>
+        <Brush />
+      </IconButton>
+    </div>
+  );
+});
+
+const SortableSpriteThumbnail = SortableElement(
+  ({ sprite, project, resourcesLoader, selected, onSelect, onContextMenu }) => {
+    return (
+      <ImageThumbnail
+        selectable
+        selected={selected}
+        onSelect={onSelect}
+        onContextMenu={onContextMenu}
+        resourceName={sprite.getImageName()}
+        resourcesLoader={resourcesLoader}
+        project={project}
+        style={styles.thumbnailExtraStyle}
+      />
+    );
+  }
+);
+
+const SortableList = SortableContainer(
+  ({
+    direction,
+    project,
+    resourcesLoader,
+    onAddSprite,
+    onEditSprites,
+    selectedSprites,
+    onSelectSprite,
+    onSpriteContextMenu,
+  }) => {
+    const spritesCount = direction.getSpritesCount();
+    return (
+      <div style={styles.spritesList}>
+        {[
+          ...mapFor(0, spritesCount, i => {
+            const sprite = direction.getSprite(i);
+            return (
+              <SortableSpriteThumbnail
+                sprite={sprite}
+                key={sprite.ptr}
+                index={i}
+                selected={!!selectedSprites[sprite.ptr]}
+                onContextMenu={(x, y) => onSpriteContextMenu(x, y, sprite)}
+                onSelect={selected => onSelectSprite(sprite, selected)}
+                resourcesLoader={resourcesLoader}
+                project={project}
+              />
+            );
+          }),
+          <AddSpriteButton
+            displayHint={!direction.getSpritesCount()}
+            key="add-sprite-button"
+            disabled
+            index={spritesCount}
+            onAdd={onAddSprite}
+            onEdit={onEditSprites}
+          />,
+        ]}
+      </div>
+    );
+  }
+);
+
+export default class SpritesList extends Component {
+  componentDidMount() {
+    if(!ipcRenderer){return};
+    ipcRenderer.on('piskelSavedChanges', (event, piskelFramePaths) => {
+      const { direction, project } = editedAnimationProp.props;
+      const resourcesManager = project.getResourcesManager();
+      direction.removeAllSprites(); /// clear the old sprite list
+      var i = 0; /// ...We need to recreate it in order to account for any new/removal/reorder frame changes made in piskel

```

Comment:
better: use `let` instead of `var`, which is the old Javascript way of declaring variable but quite unsafe as not "scoped".
even better: put the let inside the `for` (`for (let i = 0;...`)
even better: use `forEach`:

```javascript
piskelFramePaths.forEach(imagePath => {
   const imageResource = new gd.ImageResource();
   ...
});
```

Use `forEach` or `map` is prefered over a loop as there is no risk of making a mistake while writing the `for`, you don't have to declare `i`, it better conveys the meaning of what you want to do (with `forEach`, you do something for each element).
Also, `map` is to be prefered when you want to transform an array:

`const newArray = oldArray.map(element => doSomething(element))`. (which is not the case here)

Reply:
Ah Thank you. I am learning a lot =)